shiny/dbsync/app.R

```diff
@@ -0,0 +1,242 @@
+            url <- sub("sync.log", "bety.tar.gz", res$url)
+            lines <- strsplit(rawToChar(res$content), '\n', fixed = TRUE)[[1]]
+            data <- list()
+            for (line in lines) {
+                pieces <- strsplit(trimws(line), ' ', fixed=TRUE)[[1]]
+                if (length(pieces) == 8) {
+                    if (pieces[8] == 0) {
+                        servers[servers$sync_url == url, paste0('server_', pieces[7])] <<- paste(pieces[1:6], collapse = " ")
+                    }
+                } else {
+                    print(line)
+                }
+            }
+        }
+        progress$inc(amount = 1)
+    }
+    urls <- sapply(servers[,'sync_url'], function(x) { sub("bety.tar.gz", "sync.log", x) })
+    lapply(urls, function(x) { curl::curl_fetch_multi(x, done = server_log, fail = failure, handle = curl::new_handle(connecttimeout=1)) })
+    
+    # run queries in parallel
+    curl::multi_run()
+    myservers <<- servers
+    return(servers)
+}
+
+# return vector to use in polylines
+check_sync <- function(servers, what) {
+    ids <- servers$sync_host_id
+    
+    # helper function to see if two servers are connected    
+    connected <- function(src, dst) {
+        v <- servers[servers$sync_host_id==src, paste0("server_", dst)]
+        !is.null(v) && !is.na(v)
+    }
+    
+    # build up list of all connections
+    result <- c()
+    for (src in ids) {
+        for (dst in ids) {
+            if (connected(src, dst)) {
+                if (what == "color") {
+                    when <- strptime(servers[servers$sync_host_id==src, paste0("server_", dst)],
+                                     format="%a %b %d %T UTC %Y",
+                                     tz="UTC")
+                    now <- as.POSIXlt(Sys.time(), tz="UTC")
+                    tdiff <- difftime(now, when, units = "hours")
+                    if (tdiff < 0) {
+                        result <- c(result, "purple")
+                    } else if (tdiff < 6) {
+                        result <- c(result, "green")
+                    } else if (tdiff < 24) {
+                        result <- c(result, "yellow")
+                    } else {
+                        result <- c(result, "red")
+                    }
+                    mycolors <<- result
+                } else {
+                    src_x <- servers[servers$sync_host_id==src, what]
+                    dst_x <- servers[servers$sync_host_id==dst, what]
+                    result <- c(result, c(src_x, src_x + (dst_x - src_x) / 2, NA))
+                }
+            }
+        }
+    }
+    
+    # need to have at least one polyline, will draw a line from server 1 to server 1
+    if (length(result) == 0) {
+        if (what == "color") {
+            result <- c("black")
+        } else {
+            result <- c(servers[1, what], servers[1, what], NA)
+        }
+    }
+    return(result)
+}
+
+# Define UI for application that draws a histogram
```

Comment:
possible left over comment from a previous app?

Reply:
no, problem is that if there are no lines the system crashes (or I need to put an if around the code below). Figured a solution is to draw a 0 pixel line if no other lines are drawn.